File: wrappers/perl/scripts/safe_check.pl
#!/usr/bin/env perl

=head1 NAME

safe_check.pl - Minimal self-test for safe-run contract compliance

=head1 SYNOPSIS

  # Run from repository root or any location with scripts/perl/safe-run.pl
  safe_check.pl

=head1 DESCRIPTION

This script provides a minimal self-test to verify that the Perl safe-run
wrapper correctly implements the M0 conformance contract. It validates the
core contract guarantees through integration testing.

The script performs the following verification tests:

=over 4

=item 1. B<Success produces no artifacts>

Executes a command that exits with code 0 and verifies that no failure
log files are created in the FAIL-LOGS directory.

=item 2. B<Failure produces fail log>

Executes a command that exits with non-zero code (37) and verifies that
exactly one failure log file is created.

=item 3. B<Exit codes are preserved>

Verifies that the wrapper preserves both success (0) and failure (37)
exit codes from the child process.

=back

=head1 TEST ENVIRONMENT

The script creates a temporary test environment:

=over 4

=item * Creates C<.agent/FAIL-LOGS> directory in current working directory

=item * Expects C<scripts/perl/safe-run.pl> to exist relative to CWD

=item * Cleans up any existing logs before running tests

=back

=head1 ENVIRONMENT VARIABLES

=over 4

=item B<SAFE_RUN_BIN>

Optional override for safe-run binary location. If not set, the safe-run.pl
wrapper will discover the binary using its standard search order.

=item B<TMPDIR>

Temporary directory location (system default if not set). Used for creating
test environments.

=back

=head1 EXIT CODES

=over 4

=item B<0>

All contract tests passed. The safe-run wrapper is functioning correctly.

=item B<1>

Contract violation detected. Error message printed to stderr explains
which test failed and what went wrong.

=back

=head1 CONTRACT VALIDATION

This test validates the following M0 specification requirements:

=over 4

=item * M0-P1-I1: Success (exit 0) produces no failure artifacts

=item * M0-P1-I2: Failure (exit non-zero) creates exactly one log file

=item * M0-P1-I3: Exit codes are preserved exactly from child process

=back

For complete contract specification, see:

=over 4

=item * docs/conformance-contract.md - Output format and behavior contract

=item * conformance/vectors.json - Complete conformance test vectors

=item * RFC-Shared-Agent-Scaffolding-v0.1.0.md - Full specification

=back

=head1 SIDE EFFECTS

=over 4

=item * Creates C<.agent/> directory in current working directory

=item * Creates C<.agent/FAIL-LOGS/> directory

=item * May leave failure log files from test execution

=item * Removes entire C<.agent/> tree at start for clean test environment

=back

=head1 DIAGNOSTICS

Error messages follow the format:

  safe_check.pl: <error description>

Common error scenarios:

=over 4

=item B<missing scripts/perl/safe-run.pl>

The safe-run.pl wrapper could not be found at the expected relative path.
Run this script from the repository root or ensure the wrapper exists.

=item B<expected exit 0 on success, got N>

The wrapper did not preserve the success exit code. This indicates a bug
in exit code forwarding.

=item B<success should not create fail logs>

The wrapper created failure log artifacts when the command succeeded.
This violates the M0 contract requirement that success produces no artifacts.

=item B<expected exit 37 on failure, got N>

The wrapper did not preserve the failure exit code (37). This indicates
a bug in exit code forwarding.

=item B<failure should create exactly one fail log>

The wrapper either created no log file or created multiple log files
when the command failed. The contract requires exactly one log per failure.

=back

=head1 EXAMPLES

=head2 Basic Usage

  $ cd /path/to/repo
  $ perl scripts/perl/safe_check.pl
  safe_check.pl: OK:
  $ echo $?
  0

=head2 Failure Case

  $ perl scripts/perl/safe_check.pl
  safe_check.pl: expected exit 0 on success, got 1
  $ echo $?
  1

=head1 TESTING STRATEGY

This script uses a "black box" integration testing approach:

=over 4

=item 1. Invoke the actual safe-run.pl wrapper (not a mock)

=item 2. Execute real Perl commands as child processes

=item 3. Verify observable behavior (exit codes, filesystem artifacts)

=item 4. No inspection of log file contents (handled by conformance tests)

=back

For comprehensive conformance testing including log format validation,
use the conformance test harness with vectors.json.

=head1 SEE ALSO

=over 4

=item * L<safe-run.pl> - The wrapper being tested

=item * L<safe-archive.pl> - Archive failure logs

=item * conformance/vectors.json - Complete test vectors

=back

=head1 AUTHOR

RFC-Shared-Agent-Scaffolding Project

=head1 LICENSE

Unlicense - See LICENSE file in repository root

=cut

use strict;
use warnings;

use File::Path qw(make_path remove_tree);
use File::Spec;
use Cwd qw(abs_path);

sub die_bad { print STDERR "safe_check.pl: $_[0]\n"; exit 1; }

my $root = File::Spec->catdir('.agent');
my $logs = File::Spec->catdir($root, 'FAIL-LOGS');

# Start clean
remove_tree($root, { error => \my $err });
make_path($logs);

my $safe_run = File::Spec->catfile('scripts','perl','safe_run.pl');
-e $safe_run or die_bad("missing $safe_run");

# 1) Success path -> no artifacts created
{
  my @before = glob(File::Spec->catfile($logs, '*'));
  my $rc = system($^X, $safe_run, '--', $^X, '-e', 'print "ok\n"; exit 0');
  my $exit = ($rc >> 8);
  $exit == 0 or die_bad("expected exit 0 on success, got $exit");
  my @after = glob(File::Spec->catfile($logs, '*'));
  scalar(@after) == scalar(@before) or die_bad('success should not create fail logs');
}

# 2) Failure path -> creates fail log + preserves exit code
{
  my @before = glob(File::Spec->catfile($logs, '*'));
  my $rc = system($^X, $safe_run, '--', $^X, '-e', 'print "nope\n"; exit 37');
  my $exit = ($rc >> 8);
  $exit == 37 or die_bad("expected exit 37 on failure, got $exit");
  my @after = glob(File::Spec->catfile($logs, '*'));
  scalar(@after) == scalar(@before) + 1 or die_bad('failure should create exactly one fail log');
}

print "safe_check.pl: OK:\n";
exit 0;

File: wrappers/python3/scripts/safe_check.py
#!/usr/bin/env python3
"""Contract verification suite for Python3 safe-run and safe-archive scripts.

This module validates that the Python3 implementations of safe-run and
safe-archive conform to the M0 specification contracts for failure logging,
exit code preservation, and archival semantics.

:Purpose:
Provides automated contract verification to ensure:
- safe-run creates failure artifacts on non-zero exit codes
- safe-run does NOT create artifacts on successful (exit 0) execution
- safe-run preserves the wrapped command's exit code exactly
- safe-archive moves files to the archive directory
- safe-archive respects no-clobber semantics (M0-P1-I3)

:Environment Variables:
SAFE_LOG_DIR : str, optional
    Directory for failure logs (default: .agent/FAIL-LOGS)
    Used by safe-run when creating failure artifacts

:Verification Tests:
1. **Failure path**: Run command that exits with code 42
   - Verify safe-run returns exit code 42 (exit code preservation)
   - Verify exactly one new artifact created in FAIL-LOGS

2. **Success path**: Run command that exits with code 0
   - Verify safe-run returns exit code 0
   - Verify NO artifacts created in FAIL-LOGS

3. **Archive move**: Archive newest failure log
   - Verify safe-archive moves file to FAIL-ARCHIVE
   - Verify source file no longer exists (move, not copy)
   - Verify destination file exists with correct content

4. **No-clobber**: Archive file when destination already exists
   - Verify safe-archive succeeds (auto-suffix mode, M0-P1-I3)
   - Verify original archive file unchanged (no clobber)
   - Verify new file uses .2 suffix

:CLI Usage:
    python3 scripts/python3/safe_check.py

No arguments accepted. The script discovers and tests the sibling scripts
(safe_run.py, safe_archive.py) in the same directory tree.

:Exit Codes:
0
    All contract verification tests passed
1
    Contract verification failed (assertion error, missing file, etc.)
2
    Usage error (unexpected arguments provided)

:Side Effects:
- Creates temporary directories: .agent/FAIL-LOGS, .agent/FAIL-ARCHIVE
- Generates test failure logs during verification
- Moves test artifacts to archive during verification
- All side effects occur in current working directory

:Examples:
Run contract verification::

    cd /path/to/repo
    python3 scripts/python3/safe_check.py
    # Output: INFO: SAFE-CHECK: contract verification PASSED

:Contract References:
This script verifies conformance with:
- safe-run-001: Exit code preservation
- safe-run-002: Stdout/stderr capture on failure
- safe-run-003: Failure artifact generation
- M0-P1-I3: No-clobber archival semantics (auto-suffix mode)

:See Also:
- scripts/python3/safe_run.py: Wrapper under test
- scripts/python3/safe_archive.py: Archival tool under test
- docs/rust-canonical-tool.md: Canonical contract specification
"""
import os
import sys
import glob
import subprocess
from pathlib import Path
from typing import List


def eprint(*args: object) -> None:
    """Print to stderr for status messages and errors.

    :param args: Variable arguments to print (passed to print())
    :returns: None

    All output is sent to stderr to avoid interfering with stdout,
    which may be captured or redirected during testing.
    """
    print(*args, file=sys.stderr)


def die(msg: str, rc: int = 1) -> None:
    """Print error message to stderr and exit immediately.

    :param msg: Error message to print (prefixed with "ERROR: ")
    :param rc: Exit code (default: 1)
    :raises SystemExit: Always raises to terminate the program

    This is a convenience function for fatal errors during verification.
    """
    eprint(f"ERROR: {msg}")
    raise SystemExit(rc)


def usage() -> None:
    """Print usage message to stderr and exit with code 2.

    :raises SystemExit: Always exits with code 2 (usage error)
    """
    eprint("Usage: scripts/python3/safe_check.py")
    raise SystemExit(2)


def count_files(d: str) -> int:
    """Count regular files in a directory (non-recursive).

    :param d: Directory path to count files in
    :returns: Number of regular files (not directories or symlinks)

    Used to verify that safe-run creates exactly one artifact per failure
    and no artifacts on success.
    """
    return len([p for p in glob.glob(os.path.join(d, "*")) if os.path.isfile(p)])


def main(argv: List[str]) -> int:
    """Execute contract verification tests.

    :param argv: Command-line arguments (expected to be empty)
    :returns: Exit code (0 on success, 1 on verification failure, 2 on usage error)
    :raises SystemExit: Via die() or usage() on fatal errors

    Test Sequence
    -------------
    1. Setup: Create .agent/FAIL-LOGS and .agent/FAIL-ARCHIVE directories
    2. Verify scripts exist: safe_run.py, safe_archive.py
    3. Test failure path: Command exits 42, verify artifact created and exit code preserved
    4. Test success path: Command exits 0, verify no artifact created
    5. Test archive move: Archive newest log, verify move semantics
    6. Test no-clobber: Archive when destination exists, verify auto-suffix (M0-P1-I3)

    Side Effects
    ------------
    - Creates .agent/FAIL-LOGS/ directory in current working directory
    - Creates .agent/FAIL-ARCHIVE/ directory in current working directory
    - Generates test failure logs during verification
    - Moves test artifacts during archive verification
    - Prints status messages to stderr (INFO/ERROR)

    Examples
    --------
    >>> main([])  # No arguments, runs all tests
    # INFO: safe_run failure-path OK
    # INFO: safe_run success-path OK
    # INFO: safe_archive move OK
    # INFO: safe_archive no-clobber OK
    # INFO: SAFE-CHECK: contract verification PASSED
    0
    """
    if argv:
        usage()

    log_dir = os.environ.get("SAFE_LOG_DIR", ".agent/FAIL-LOGS")
    os.makedirs(log_dir, exist_ok=True)
    os.makedirs(".agent/FAIL-ARCHIVE", exist_ok=True)

    if not Path("scripts/python3/safe_run.py").is_file():
        die("Missing scripts/python3/safe_run.py")
    if not Path("scripts/python3/safe_archive.py").is_file():
        die("Missing scripts/python3/safe_archive.py")

    before = count_files(log_dir)

    # failure path
    rc = subprocess.call(
        [
            sys.executable,
            "scripts/python3/safe_run.py",
            "--",
            sys.executable,
            "-c",
            'import sys; print("hello"); print("boom", file=sys.stderr); raise SystemExit(42)',
        ],
        stdout=subprocess.DEVNULL,
    )
    if rc != 42:
        die(f"safe_run did not preserve exit code (expected 42, got {rc})")
    after = count_files(log_dir)
    if after != before + 1:
        die(f"safe_run failure did not create exactly one artifact (before={before} after={after})")
    eprint("INFO: safe_run failure-path OK")

    # success path
    before = after
    rc = subprocess.call(
        [
            sys.executable,
            "scripts/python3/safe_run.py",
            "--",
            sys.executable,
            "-c",
            'print("ok"); raise SystemExit(0)',
        ],
        stdout=subprocess.DEVNULL,
    )
    if rc != 0:
        die(f"safe_run success returned non-zero ({rc})")
    after = count_files(log_dir)
    if after != before:
        die(f"safe_run success created artifacts (before={before} after={after})")
    eprint("INFO: safe_run success-path OK")

    # archive newest
    files = sorted(
        [p for p in glob.glob(os.path.join(log_dir, "*")) if os.path.isfile(p)],
        key=os.path.getmtime,
        reverse=True,
    )
    if not files:
        die("No fail logs found to test archiving")
    newest = files[0]
    base = os.path.basename(newest)
    dest = os.path.join(".agent/FAIL-ARCHIVE", base)

    rc = subprocess.call(
        [sys.executable, "scripts/python3/safe_archive.py", newest],
        stdout=subprocess.DEVNULL,
    )
    if rc != 0:
        die(f"safe_archive failed ({rc})")
    if not os.path.exists(dest):
        die(f"Archive file missing: {dest}")
    if os.path.exists(newest):
        die("Source file still exists (expected moved)")
    eprint("INFO: safe_archive move OK")

    # no-clobber
    dummy = os.path.join(log_dir, base)
    with open(dummy, "w", encoding="utf-8") as fh:
        fh.write("dummy\n")
    rc = subprocess.call(
        [sys.executable, "scripts/python3/safe_archive.py", dummy],
        stdout=subprocess.DEVNULL,
    )
    if rc != 0:
        die(f"safe_archive no-clobber failed ({rc})")
    with open(dest, "r", encoding="utf-8", errors="replace") as fh:
        contents = fh.read()
    if "hello" not in contents:
        die("Archive content changed unexpectedly (no-clobber violation suspected)")
    eprint("INFO: safe_archive no-clobber OK")

    eprint("INFO: SAFE-CHECK: contract verification PASSED")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
